website_sale_osc/controllers/main.py

```diff
@@ -0,0 +1,276 @@
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from openerp import SUPERUSER_ID
+from openerp.addons.website_sale.controllers.main import website_sale
+from openerp.addons.web import http
+from openerp.addons.web.http import request
+from openerp.report import report_sxw
+
+
+class website_sale(website_sale):
+
+    """Add aditional functions to the website_sale controller."""
+
+    mandatory_billing_fields = ["name", "phone", "email", "street", "city",
+                                "country_id", "zip"]
+    optional_billing_fields = ["street2", "state_id", "vat", "vat_subjected"]
+
+    @http.route(['/shop/checkout'], type='http', auth='public',
+                website=True, multilang=True)
+    def checkout(self, **post):
+        """Checkout controller."""
+        # if onestepcheckout is deactivated use the normal checkout
+        if not request.website.use_osc:
+            return super(website_sale, self).checkout()
+
+        # must have a draft sale order with lines at this point
+        order = request.website.sale_get_order()
+
+        redirection = self.checkout_redirection(order)
+        if redirection:
+            return redirection
+
+        values = self.checkout_values(post)
+
+        partner = request.env['res.users'].sudo().browse(
+            request.uid).partner_id
+
+        # get countries dependent on website settings
+        countries_domain = []
+        if not request.website.use_all_checkout_countries:
+            countries_domain = [('id', 'in',
+                                 request.website.checkout_country_ids.ids)]
+
+        values['countries'] = request.env['res.country'].search(
+            countries_domain)
+
+        if not post and request.uid != request.website.user_id.id:
+            values['checkout'].update({'street': partner.street_name,
+                                       'street_number': partner.street_number})
+
+        result = self.payment(post=post)
+        values.update(result.qcontext)
+
+        # get additional tax information
+        values['tax_overview'] = request.env['sale.order'].tax_overview(order)
+        return request.website.render(
+            'website_sale_osc.osc_onestepcheckout', values)
+
+    @http.route(['/shop/checkout/confirm_address/'], type='json',
+                auth='public', website=True, multilang=True)
+    def confirm_address(self, **post):
```

Comment:
Can't you use some code of the super? This huge amount of code makes not very maintainable along time